docs/proposals/16-adminkubeconfig-subresource.md

```diff
@@ -0,0 +1,126 @@
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+  - [Alternatives](#alternatives)
+
+## Summary
+
+This `GEP` introduces new  `Shoot` subresource called `AdminKubeConfigRequest` allowing for users to dynamically generate a short-lived `kubeconfig` that can be used to access the `Shoot` cluster as `cluster-admin`.
+
+## Motivation
+
+Today, when access to the created `Shoot` clusters is needed, a `kubeconfig` with static token credentials is used. This static token is in the `system:masters` group, granting it `cluster-admin` privileges. The `kubeconfig` is generated when the cluster is reconciled, stored in `ShootState` and replicated in the `Project`'s namespace in a `Secret`. End-users can fetch the secret and use the `kubeconfig` inside it.
+
+There are several problems with this approach:
+
+- The token in the `kubeconfig` does not have any expiration, so end-users have to request a `kubeconfig` credential rotation if they want revoke the token.
+- There is no user identity in the token. e.g. if user `Joe` gets the `kubeconfig` from the `Secret`, user in that token would be `system:cluster-admin` and not `Joe` when accessing the `Shoot` cluster with it. This makes auditing events in the cluster almost impossible.
+
+### Goals
+
+- Add a `Shoot` subresource called `adminkubeconfig` that would produce a `kubeconfig` used to access that `Shoot` cluster.
+- The `kubeconfig` is not stored in the API Server, but generated for each request.
+- In the `AdminKubeConfigRequest` send to that subresource, end-users can specify the expiration time of the credential.
+- The identity (user) in the Gardener cluster would be part of the identity (x509 client certificate). E.g if `Joe` authenticates against the Gardener API server, the generated certificate for `Shoot` authentication would have the following subject:
+
+  - Common Name: `Joe`
+  - Organisation: `system:masters`
+
+- The maximum validity of the certificate can be enforced by setting a flag on the `gardener-apiserver`.
+- Deprecate and remove the old `{shoot-name}.kubeconfig` secrets in each `Project` namespace.
+
+### Non-Goals
+
+- Generate `OpenID Connect` kubeconfigs
+
+## Proposal
+
+The `gardener-apiserver` would serve a new `shoots/adminkubeconfig` resource. It can only accept `CREATE` calls and accept `AdminKubeConfigRequest`. A `AdminKubeConfigRequest` would have the following structure:
+
+```yaml
+apiVersion: authentication.gardener.cloud/v1alpha1
+kind: AdminKubeConfigRequest
+spec:
+  expirationSeconds: 3600
+```
+
+Where `expirationSeconds` is the validity of the certificate in seconds. It this case it would be `1 hour`. The maximum validity of a `AdminKubeConfigRequest` is configured by `--shoot-admin-kubeconfig-max-token-expiration` flag in the `gardener-apiserver`.
```

Comment:
```suggestion
Where `expirationSeconds` is the validity of the certificate in seconds. It this case it would be `1 hour`. The maximum validity of a `AdminKubeConfigRequest` is configured by `--shoot-admin-kubeconfig-max-certificate-duration` flag in the `gardener-apiserver`.
```

Reply:
I'll rather keep it `--shoot-admin-kubeconfig-max-expiration`

In the future we might change the implementation to not use `certificates`, but some other authentication methods.

Reply:
OK, makes sense.